jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.c

```diff
@@ -525,19 +619,33 @@ ecma_builtin_string_prototype_object_replace (ecma_value_t this_value, /**< this
         else
         {
           replace_ctx.matched_p = curr_p;
-          replace_ctx.match_byte_pos = byte_offset;
+          replace_ctx.match_byte_pos = (lit_utf8_size_t) (curr_p - replace_ctx.string_p);
 
           ecma_builtin_replace_substitute (&replace_ctx);
         }
 
-        const lit_utf8_byte_t *const match_end_p = curr_p + search_size;
-        ecma_stringbuilder_append_raw (&replace_ctx.builder,
-                                       match_end_p,
-                                       (lit_utf8_size_t) (input_end_p - match_end_p));
-        result_string_p = ecma_stringbuilder_finalize (&replace_ctx.builder);
-        break;
+        if (!replace_all
+            || last_match_end_p == input_end_p)
+        {
+          break;
+        }
+
+        if (search_size != 0)
+        {
+          curr_p = last_match_end_p;
```

Comment:
`pos` also needs to be incremented by the length of the search string, otherwise following matches will use an incorrect position. This should fix the last failing test.

The `ecma_string_get_chars` call on line 557 could be updated to retrieve the length.